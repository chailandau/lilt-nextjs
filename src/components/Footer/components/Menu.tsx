import { FC } from 'react';

import { FooterProps } from '../Footer';

import Link from '@/atoms/Link';
import { getLink } from '@/utils/getLink';

const Menu: FC<FooterProps> = ({ menuItems }) => {
    const menuContent = menuItems?.map((menuItem) => {
        if (
            !menuItem.linkType ||
            (menuItem.linkType === 'external' && !menuItem.label) ||
            (menuItem.linkType === 'internal' &&
                (!menuItem?.internalLink?.slug ||
                    !menuItem.internalLink?.title))
        ) {
            return null;
        }

        const menuLink = getLink({
            linkType: menuItem?.linkType,
            externalLink: menuItem?.externalLink,
            internalLink: menuItem?.internalLink
        });

        const menuLabel =
            menuItem?.linkType === 'external'
                ? menuItem?.label
                : menuItem?.internalLink?.title;

        return (
            <li key={menuItem?.id}>
                <Link key={menuItem?.id} href={menuLink}>
                    {menuLabel}
                </Link>
            </li>
        );
    });

    return <ul>{menuContent}</ul>;
};

export default Menu;
