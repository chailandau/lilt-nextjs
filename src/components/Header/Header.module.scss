.header {
    flex-direction: row !important;
    padding-top: rem(24);
    padding-bottom: rem(16);
    margin-bottom: rem(-16);
    position: fixed;
    top: 0;
    z-index: 99999;
    width: 100vw;
    @include transition(padding-top, padding-bottom);
    @include paddingX(0);
    &:before {
        display: block;
        content: '';
        position: absolute;
        width: 100%;
        height: 100%;
        top: 0;
        left: 0;
        z-index: 101;
        background: color(dark, blue);
    }
    &__content {
        flex-direction: row;
        gap: rem(40);
        justify-content: space-between;
        align-items: center;
    }
    .logo,
    .right-content {
        z-index: 101;
    }
    .logo {
        width: rem(250);
        padding: unset;
        @include transition(width);
        &:before {
            display: none;
        }
    }
    .right-content {
        align-items: center;
        gap: rem(24);
        @include transition(gap);
        @include laptop-lg {
            gap: rem(32);
        }
        @include desktop-sm {
            gap: rem(44);
        }
        .tablet-btn {
            display: none;
            text-align: center;
            @include transition(padding, background-position);

            @include tablet {
                display: block;
            }
        }
    }
    &__sticky {
        @include paddingY(4);
        @include tablet {
            @include paddingY(12);
        }
        &:before {
            box-shadow: shadow(lightFuchsia);
            border-bottom: 1px solid color(neutral, grayBlue);
        }
        .logo {
            width: rem(180);
        }
        .tablet-btn {
            padding: rem(10) rem(15);
        }
        &.header__menu-open {
            padding-top: rem(24);
            padding-bottom: rem(16);
            &:before {
                display: none;
            }
            .logo {
                width: rem(250);
            }
        }
    }
}
