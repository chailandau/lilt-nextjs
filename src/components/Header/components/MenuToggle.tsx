import classNames from 'classnames';

import styles from './MenuToggle.module.scss';

import Container from '@/atoms/Container';
import useStore from '@/store/useStore';

const MenuToggle = () => {
    const { menuOpen, setMenuOpen, setOpenSubmenu } = useStore();

    const handleClick = () => {
        setMenuOpen(!menuOpen);
        setOpenSubmenu(null);
    };

    const classList = classNames(
        styles['menu-toggle'],
        menuOpen && styles['open']
    );

    return (
        <Container
            as='button'
            className={classList}
            onClick={handleClick}
            ariaLabel={'menu toggle'}
        >
            <Container className={styles['menu-toggle-container']}>
                <Container className={styles['menu-toggle-inner']}></Container>
            </Container>
        </Container>
    );
};

export default MenuToggle;
