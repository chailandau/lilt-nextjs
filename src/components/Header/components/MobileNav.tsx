import { m } from 'framer-motion';

import styles from './MobileNav.module.scss';

import Link from '@/atoms/Link/Link';
import Button from '@/molecules/Button/Button';
import Flex from '@/molecules/Flex/Flex';
import useStore from '@/store/useStore';
import { menuAnimations } from '@/utils/framer/animations';
import LazyAnimatePresence from '@/utils/framer/LazyAnimatePresence';

const MobileNav = () => {
    const menuOpen = useStore((state) => state.menuOpen);

    return (
        <LazyAnimatePresence>
            {menuOpen &&
                <m.nav
                    className={styles['mobile-nav']}
                    variants={menuAnimations}
                    animate={menuOpen ? 'open' : 'closed'}
                    initial="closed"
                    exit="closed"
                >
                    <ul>
                        <li><Link className={styles['link']} href='http://localhost:6006/home'>Booking</Link></li>
                        <li><Link className={styles['link']} href='http://localhost:6006/'>Schools & Camps</Link></li>
                        <li><Link className={styles['link']} href='http://localhost:6006/'>Covid-19 Guidelines</Link></li>
                    </ul>
                    <Flex className={styles['button-wrapper']}>
                        <Button link='http://localhost:6006/'>Get in touch</Button>
                    </Flex>
                </m.nav>
            }

        </LazyAnimatePresence>
    );
};

export default MobileNav;
