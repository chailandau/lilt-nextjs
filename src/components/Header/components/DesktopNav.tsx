import { FC } from 'react';

import { HeaderProps } from '../Header';

/**
 * Flex (and its styles) must be imported before
 * current styles for proper Storybook rendering
 */
// eslint-disable-next-line import/order
import Flex from '@/molecules/Flex/Flex';
import styles from './DesktopNav.module.scss';
import Menu from './Menu';

import LazyMotionDom from '@/utils/framer/LazyMotionDom';

const DesktopNav: FC<HeaderProps> = ({ menuItems }) => (
    <LazyMotionDom>
        <Flex as='nav' className={styles['desktop-nav']}>
            <Menu menuItems={menuItems} />
        </Flex>
    </LazyMotionDom>
);

export default DesktopNav;
