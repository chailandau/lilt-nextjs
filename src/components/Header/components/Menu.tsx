import { FC, KeyboardEvent } from 'react';

import { HeaderProps } from '../Header';

import styles from './Menu.module.scss';
import Submenu from './Submenu';

import Link from '@/atoms/Link';
import useStore from '@/store/useStore';
import { getLink } from '@/utils/getLink';

export interface MenuProps extends HeaderProps {
    /** Optional classname */
    className?: string;
}

const Menu: FC<MenuProps> = ({ menuItems, className = styles['menu'] }) => {
    const menuContent = menuItems?.map((menuItem, index) => {
        if (
            !menuItem.linkType ||
            (menuItem.linkType === 'external' && !menuItem.label) ||
            (menuItem.linkType === 'internal' &&
                (!menuItem?.internalLink?.slug ||
                    !menuItem.internalLink?.title))
        ) {
            return null;
        }
        const { setOpenSubmenu } = useStore();

        const handleKeyDown = (
            e: KeyboardEvent<HTMLAnchorElement | HTMLButtonElement>
        ) => {
            const isLastItem = index === menuItems?.length - 1;
            const isSubmenu = menuItem?.submenuItems === undefined;
            if (e.key === 'Tab' && !e.shiftKey && isLastItem && isSubmenu) {
                setOpenSubmenu(null);
            }
        };

        const menuLink = getLink({
            linkType: menuItem?.linkType,
            externalLink: menuItem?.externalLink,
            internalLink: menuItem?.internalLink
        });

        const overridePageName =
            menuItem?.overridePageName && menuItem?.internalCustomLabel;

        const internalMenuLabel = overridePageName
            ? menuItem?.internalCustomLabel
            : menuItem?.internalLink?.title;

        const menuLabel =
            menuItem?.linkType === 'external'
                ? menuItem?.label
                : internalMenuLabel;

        if (menuItem?.linkType === 'submenu') {
            return (
                menuItem?.label &&
                menuItem?.submenuItems &&
                menuItem.submenuItems.length > 0 && (
                    <Submenu key={menuItem?.id} menuItem={menuItem} />
                )
            );
        } else {
            return (
                <li key={menuItem?.id}>
                    <Link
                        key={menuItem?.id}
                        href={menuLink || ''}
                        className={styles['menu-link']}
                        onKeyDown={(e) => handleKeyDown(e)}
                    >
                        {menuLabel}
                    </Link>
                </li>
            );
        }
    });

    return <ul className={className}>{menuContent}</ul>;
};

export default Menu;
