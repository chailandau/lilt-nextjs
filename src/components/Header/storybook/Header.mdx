import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as HeaderStories from './Header.stories';

<Meta of={HeaderStories} />

# Header Component

<Canvas of={HeaderStories.Default} />

The `Header` component accepts a `menuItems` array and a `callToAction` object.

## PayloadCMS

From dashboard, navigate to `Global` > `Header`.

### menuItems

Menu items contain the following fields (these vary depending on which `Menu Item Type` is selected):

-   `Menu Item Type` - select whether menu item is an internal/external
    link or a submenu
-   `Internal Link` - select site page to link to
    -   `Override Page Name` - override site page name with custom label
    -   `Custom Label` - input internal menu item's custom label
-   `External Link` - input a valid external URL - `Label` - menu item's label
-   `Submenu` - create nested menu item.
    -   Contains same options as regular menu item, except for submenu option (one level of submenu is currently supported).

<img
    src='docs/header/menu-items.gif'
    alt='walkthrough of how to set header menu items in PayloadCMS'
/>

<hr />

### callToAction

    - `Call To Action` - select the header's call to action button via the dropdown
        - Content editors can select an existing CTA or create a new one by clicking the `+` sign to the right of the dropdown.
        - See `Button` component [docs](?path=/docs/components-button--docs#buttondata) for a walkthrough on creating a Button in PayloadCMS.

<hr />

<img
    src='docs/header/cta.gif'
    alt='walkthrough of how to set header call to action in PayloadCMS'
/>

<br />

### Props Table

<ArgTypes />
