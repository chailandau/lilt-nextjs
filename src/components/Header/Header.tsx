'use client';

import FocusTrap from 'focus-trap-react';
import { FC, useEffect } from 'react';

import Button from '../Button/Button';

import DesktopNav from './components/DesktopNav';
import MenuToggle from './components/MenuToggle';
import MobileNav from './components/MobileNav';
import styles from './Header.module.scss';

import { CallToAction, Header_MenuItems } from '@/api/graphqlTypes';
import Logo from '@/assets/svg/Logo.svg';
import Image from '@/atoms/Image';
import Link from '@/atoms/Link';
import Flex from '@/molecules/Flex';
import Section from '@/molecules/Section';
import useStore from '@/store/useStore';
import { laptopQuery, useMediaQuery } from '@/utils/hooks/useMediaQuery';

export interface HeaderProps {
    /* Menu items to display */
    menuItems: Header_MenuItems[] | null;
    /* CTA object */
    callToAction?: CallToAction | null;
}
const Header: FC<HeaderProps> = ({ menuItems, callToAction }) => {
    const isLaptop = useMediaQuery(laptopQuery);

    const { menuOpen, setMenuOpen } = useStore();

    useEffect(() => {
        if (isLaptop) {
            setMenuOpen(false);
        }
    }, [isLaptop]);

    return (
        <FocusTrap active={menuOpen}>
            <Section as='header' className={styles['header']}>
                <Link
                    href={process.env.NEXT_PUBLIC_BASE_URL as string}
                    className={styles['logo']}
                >
                    <Image
                        src={Logo}
                        alt='Long Island Laser Tag'
                        hasBorder={false}
                        priority
                    />
                </Link>

                <Flex className={styles['right-content']}>
                    {isLaptop && (
                        <DesktopNav
                            menuItems={menuItems}
                            callToAction={callToAction}
                        />
                    )}
                    {callToAction && (
                        <Button
                            buttonData={callToAction}
                            className={styles['tablet-btn']}
                        />
                    )}
                    <MenuToggle />
                </Flex>
                {!isLaptop && menuOpen && (
                    <MobileNav
                        menuItems={menuItems}
                        callToAction={callToAction}
                    />
                )}
            </Section>
        </FocusTrap>
    );
};

export default Header;
