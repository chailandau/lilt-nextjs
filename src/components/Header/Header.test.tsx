import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { act } from 'react-test-renderer';

import DesktopNav from './components/DesktopNav';
import Menu from './components/Menu';
import MenuToggle from './components/MenuToggle';
import MobileNav from './components/MobileNav';
import Header from './Header';

import {
    menuItemNoLabel,
    menuItemNoLinktype,
    menuItems,
    menuItemsInternalAndExternal
} from '@/__mocks__/Header.mock';
import { mockIsLaptop } from '@/__mocks__/useMediaQuery.mock';
import { testAxeAndSnapshot } from '@/utils/testHelpers';

describe('Header', () => {
    describe('Menu', () => {
        it('renders menu items correctly', () => {
            render(<Menu menuItems={menuItemsInternalAndExternal} />);

            const menuItemsElements = screen.getAllByRole('listitem');

            expect(menuItemsElements).toHaveLength(2);
        });

        it("doesn't render menu item if linkType not set", () => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            render(<Menu menuItems={menuItemNoLinktype as any} />);

            const menuItemsElements = screen
                .getByRole('list')
                .querySelector('li');

            expect(menuItemsElements).toBeNull();
        });

        it("doesn't render menu item if linkType is external + label isn't set", () => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            render(<Menu menuItems={menuItemNoLabel} />);

            const menuItemsElements = screen
                .getByRole('list')
                .querySelector('li');

            expect(menuItemsElements).toBeNull();
        });
        describe('Submenu', () => {
            let aboutMenu: HTMLElement;
            let aboutSubmenu: HTMLUListElement | null;
            let ourTechLink: Node | Window;

            beforeEach(() => {
                render(<Menu menuItems={menuItems} />);

                aboutMenu = screen.getByRole('button', {
                    name: 'About'
                });

                fireEvent.keyDown(aboutMenu, { key: 'Enter' });

                aboutSubmenu = aboutMenu.querySelector('ul');

                ourTechLink = screen.getByRole('link', {
                    name: 'Our Tech'
                });
            });

            it('exits submenu on tab forward', () => {
                fireEvent.keyDown(ourTechLink, { key: 'Tab', shiftKey: false });

                expect(aboutSubmenu).not.toBeInTheDocument();
            });
            it('exits submenu on tab backward', () => {
                fireEvent.keyDown(aboutMenu, { key: 'Tab', shiftKey: true });

                expect(aboutSubmenu).not.toBeInTheDocument();
            });
            it('exits submenu when enter is clicked again', () => {
                fireEvent.keyDown(aboutMenu, { key: 'Enter' });

                expect(aboutSubmenu).not.toBeInTheDocument();
            });
        });

        testAxeAndSnapshot({ component: <Menu menuItems={menuItems} /> });
    });
    describe('Menu Toggle', () => {
        it('opens navigation on click', async () => {
            const header = render(<Header menuItems={menuItems} />);

            const menuToggle = await header.findByLabelText('menu toggle');

            fireEvent.click(menuToggle);

            expect(menuToggle).toHaveClass('open');
        });

        testAxeAndSnapshot({ component: <MenuToggle /> });
    });
    describe('Mobile Nav', () => {
        beforeEach(async () => {
            act(() => {
                mockIsLaptop(false);
            });
            render(<Header menuItems={menuItems} />);
            fireEvent.click(await screen.findByLabelText('menu toggle'));
        });
        it('removes mobile nav when isLaptop is true', async () => {
            act(() => {
                mockIsLaptop(true);
            });

            expect(await screen.findByRole('navigation')).not.toHaveClass(
                'mobile-nav'
            );
        });
        it('opens submenu on button click', async () => {
            fireEvent.click(screen.getByRole('button', { name: 'About' }));
            screen.debug();
            expect(await screen.findByText('Our Tech')).toBeInTheDocument();
        });
        testAxeAndSnapshot({ component: <MobileNav menuItems={menuItems} /> });
    });
    describe('DesktopNav', () => {
        const DesktopNavEl = <DesktopNav menuItems={menuItems} />;
        beforeEach(() => {
            render(DesktopNavEl);
            act(() => {
                mockIsLaptop(true);
            });
        });

        it('renders correctly', () => {
            expect(screen.getByRole('navigation')).toBeInTheDocument();
        });

        it('ignores submenu button click', () => {
            fireEvent.click(screen.getByRole('button', { name: 'About' }));
            expect(screen.queryByText('Our Tech')).not.toBeInTheDocument();
        });

        it('opens and closes submenu on mouse enter and leave', async () => {
            fireEvent.mouseEnter(screen.getByRole('button', { name: 'About' }));

            await waitFor(async () =>
                expect(screen.queryByText('Our Tech')).toBeInTheDocument()
            );
            fireEvent.mouseLeave(screen.getByRole('button', { name: 'About' }));

            await waitFor(async () =>
                expect(screen.queryByText('Our Tech')).not.toBeInTheDocument()
            );
        });

        testAxeAndSnapshot({ component: DesktopNavEl });
    });
});
