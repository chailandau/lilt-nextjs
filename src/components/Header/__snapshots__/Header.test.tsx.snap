// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header DesktopNav matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="flex desktop-nav"
  >
    <ul
      class="menu"
    >
      <li>
        <a
          class="link underline-hover menu-link"
          href="/booking"
          rel=""
          target=""
        >
          Booking
        </a>
      </li>
      <li>
        <a
          class="link underline-hover menu-link"
          href="/events"
          rel=""
          target=""
        >
          Events
        </a>
      </li>
      <li>
        <a
          class="link underline-hover menu-link"
          href="/schools-camps"
          rel=""
          target=""
        >
          Schools & Camps
        </a>
      </li>
      <li>
        <button
          class="button link submenu-btn"
        >
          About
          <svg
            class="gray sm"
            style="transform: none;"
          >
            <use
              href="/icons/icon-sprites.svg#caret"
            />
          </svg>
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`Header Menu Toggle matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="menu toggle"
    class="menu-toggle open"
  >
    <div
      class="menu-toggle-container"
    >
      <div
        class="menu-toggle-inner"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`Header Menu matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="menu"
  >
    <li>
      <a
        class="link underline-hover menu-link"
        href="/booking"
        rel=""
        target=""
      >
        Booking
      </a>
    </li>
    <li>
      <a
        class="link underline-hover menu-link"
        href="/events"
        rel=""
        target=""
      >
        Events
      </a>
    </li>
    <li>
      <a
        class="link underline-hover menu-link"
        href="/schools-camps"
        rel=""
        target=""
      >
        Schools & Camps
      </a>
    </li>
    <li>
      <button
        class="button link submenu-btn"
      >
        About
        <svg
          class="gray sm"
        >
          <use
            href="/icons/icon-sprites.svg#caret"
          />
        </svg>
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Header Mobile Nav matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="mobile-nav"
    style="opacity: 0; height: 0px; transform: translateY(-100%) translateZ(0);"
  >
    <ul
      class="menu"
    >
      <li>
        <a
          class="link underline-hover menu-link"
          href="/booking"
          rel=""
          target=""
        >
          Booking
        </a>
      </li>
      <li>
        <a
          class="link underline-hover menu-link"
          href="/events"
          rel=""
          target=""
        >
          Events
        </a>
      </li>
      <li>
        <a
          class="link underline-hover menu-link"
          href="/schools-camps"
          rel=""
          target=""
        >
          Schools & Camps
        </a>
      </li>
      <li>
        <button
          class="button link submenu-btn"
        >
          About
          <svg
            class="gray sm"
            style="transform: none;"
          >
            <use
              href="/icons/icon-sprites.svg#caret"
            />
          </svg>
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
